### PR TITLE
Remove dynamo0.1 API code and tests (closes #282)

### LINKED ISSUE #282: Remove the dynamo 0.1 API
Dynamo 0.1 has been superseded by 0.3 and has not been developed for several years.

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
This is a relatively straight forward PR that removed dynamo0.1 api code. There is one place where I found what I believe to be an error in the code that reads the configuration file which I have fixed. The code was assuming that the keyword for the default api in the `DEFAULT` section of the config file was `API`, but it is `DEFAULTAPI`. This did not come to light before as this particular path was not being exercised (the test was using the logic associated with having a single api section). I now do exercise this path with this test and have changed the code accordingly. This required changes to a couple of tests as well.

**Comment by rupertford** [comment]
Ready for first review from @TeranIvy (thanks for volunteering).

File: src/psyclone/tests/generator_test.py
Comment (TeranIvy) [inline src/psyclone/tests/generator_test.py:547]: This should really be "unsupported APIs" instead of "unsupported API's" (plural, not possesive). I would kindly ask for the error message and the associated tests to be updated.

File: src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90:2]: Do we actually need this test? I thought that `-d` option is tested elsewhere.

File: src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90:4]: Is this file moved/copied or is it new? In any case, "2021" should be "2022".

File: src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90:34]: I would suggest "Author: R. W. Ford, STFC Daresbury Lab"

File: src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90:32]: I would suggest "Author: R. W. Ford, STFC Daresbury Lab" and similar below.

File: src/psyclone/tests/test_files/gocean1p0/gocean_default.cfg
Comment (TeranIvy) [inline src/psyclone/tests/test_files/gocean1p0/gocean_default.cfg:41]: As far as I can see, there is no such setting as `API` in the configuration file `psyclone.cfg` nor it is described in the documentation.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Thanks for doing this! It looks good overall, however I have some suggestions and requests (see inline).

Tests with compilation pass, examples run and the documentation builds. I can confirm that the use of `dynamo0.1` API in the code, tests, examples and tutorials is now removed.

I will merge this PR after #1598 is merged to master, so I would suggest just addressing the review comments for now and then refreshing the branch after the documentation change is merged.

File: src/psyclone/configuration.py
Comment (rupertford) [reply src/psyclone/configuration.py:270]: Agreed.

File: src/psyclone/configuration.py
Comment (rupertford) [reply src/psyclone/configuration.py:814]: OK, done (but not checked yet!)

File: src/psyclone/configuration.py
Comment (rupertford) [reply src/psyclone/configuration.py:879]: This is the Python3 way of doing it so we no longer need to use six. Notice the `raise ... from ...` format.

File: src/psyclone/configuration.py
Comment (rupertford) [reply src/psyclone/configuration.py:890]: Again this is the Python3 form

File: src/psyclone/configuration.py
Comment (rupertford) [reply src/psyclone/configuration.py:932]: This is the Python3 form

File: src/psyclone/configuration.py
Comment (rupertford) [reply src/psyclone/configuration.py:950]: This is the Python3 form

File: src/psyclone/parse/kernel.py
Comment (rupertford) [reply src/psyclone/parse/kernel.py:4]: I didn't get that error in my version of pylint. Anyway I've rewritten the line so hopefully it will no longer be a problem.

File: src/psyclone/psyGen.py
Comment (rupertford) [reply src/psyclone/psyGen.py:2573]: I don't tend to use that form but seeing as you asked and that the method below it does the same thing I've change it.

File: src/psyclone/psyGen.py
Comment (rupertford) [reply src/psyclone/psyGen.py:2575]: Oops. In general I am not a possessive person.

File: src/psyclone/tests/alggen_test.py
Comment (rupertford) [reply src/psyclone/tests/alggen_test.py:4]: 'tis done.

File: src/psyclone/tests/alggen_test.py
Comment (rupertford) [reply src/psyclone/tests/alggen_test.py:428]: OK. I've also done the same for the other comments in the file.

File: src/psyclone/tests/generator_test.py
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:580]: Agreed.

File: src/psyclone/tests/generator_test.py
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:547]: Good s'pot. I've also updated generator.py which needed copyright update and I got rid of the executable line and encoding information at the top of the file.

File: src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90
Comment (rupertford) [reply src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90:2]: I think it is right and is needed. This is used in `generator_test.py`. We are replacing the use of the `dynamo0.1` API with a `dynamo0.3` version. To test the script we need a new kernel file buried in a subdirectory of `test_files/dynamo.p3`. If it is not new then other tests fail as the same file is found more than once and an exception is raised.

File: src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90
Comment (rupertford) [reply src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90:4]: It is copied and is just there to make sure there is not a false match of a kernel when there is a hierarchy of directories. I've pretty much replicated what we did in the `dynamo0.1` tests. Updated to 2022, thanks.

File: src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90
Comment (rupertford) [reply src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90:34]: OK.

File: src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90
Comment (rupertford) [reply src/psyclone/tests/test_files/dynamo0p3/kernels/dead_end/no_really/testkern_qr.F90:32]: Updated.

File: src/psyclone/tests/test_files/gocean1p0/gocean_default.cfg
Comment (rupertford) [reply src/psyclone/tests/test_files/gocean1p0/gocean_default.cfg:41]: You're quite right. I think I left it in there as a test was expecting it then I forgot to remove it afterwards. Anyway, well spotted.

**Comment by rupertford** [comment]
Ready for next review from @TeranIvy subject to CI.

File: src/psyclone/psyGen.py
Comment (TeranIvy) [reply src/psyclone/psyGen.py]: Thanks, I asked it for the consistency with the other methods in the class.

File: src/psyclone/psyGen.py
Comment (TeranIvy) [reply src/psyclone/psyGen.py]: OK, let's not got there 🤣

File: src/psyclone/tests/alggen_test.py
Comment (TeranIvy) [reply src/psyclone/tests/alggen_test.py]: Thanks!

File: src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/1_single_invoke_kern.f90:2]: Ah, I see. Thanks for explaining!

**Review by TeranIvy — APPROVED** [review]
All well now.

Tests with compilation pass, examples run and the documentation builds. Again, I can confirm that the use of dynamo0.1 API in the code, tests, examples and tutorials is now removed.

Updated changelog and user guide. Ready to go subject to CI passing.

**Pull request merged**
